### PR TITLE
Attribute inference attack that uses membership inference attack

### PR DESCRIPTION
**Pull request by abigailgold**
# Description

This is a new type of attribute inference attack.

## Type of change

Please check all relevant options.

- [ ] Improvement (non-breaking)
- [ ] Bug fix (non-breaking)
- [X] New feature (non-breaking)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [X] This change requires a documentation update

# Testing

Please describe the tests that you ran to verify your changes. Consider listing any relevant details of your test configuration.

Ran new unit tests + existing tests for membership attacks

**Test Configuration**:
- OS: MacOS 10.15.7
- Python version: 3.7
- ART version or commit number
- TensorFlow / Keras / PyTorch / MXNet version

# Checklist

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code
- [ ] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: art/attacks/inference/attribute_inference/meminf_based.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+from art.attacks.attack import AttributeInferenceAttack, MembershipInferenceAttack
+from art.utils import check_and_transform_label_format, float_to_categorical, floats_to_one_hot
+
+if TYPE_CHECKING:
+    from art.utils import CLASSIFIER_TYPE
+
+logger = logging.getLogger(__name__)
+
+
+class AttributeInferenceUsingMembershipInference(AttributeInferenceAttack):
```
Comment (beat-buesser) [inline art/attacks/inference/attribute_inference/meminf_based.py:40]: This class name is getting quite long.

File: art/attacks/inference/attribute_inference/meminf_based.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+# The above copyright notice and this permission notice shall be included in all copies or substantial portions of the
+# Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
+# WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+# TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+"""
+This module implements attribute inference attacks.
```
Comment (beat-buesser) [inline art/attacks/inference/attribute_inference/meminf_based.py:19]:
```suggestion
This module implements attribute inference attacks using membership inference attacks.
```

File: art/attacks/inference/attribute_inference/meminf_based.py [hunk 3]
```diff
@@ -0,0 +1,158 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2020
```
Comment (beat-buesser) [inline art/attacks/inference/attribute_inference/meminf_based.py:3]:
```suggestion
# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2021
```

File: art/attacks/inference/membership_inference/black_box.py [hunk 4]
```diff
@@ -74,7 +74,7 @@ def __init__(
         :param attack_model: The attack model to train, optional. If none is provided, a default model will be created.
         """
 
-        super().__init__(estimator=classifier)
+        super().__init__(classifier=classifier)
```
Comment (beat-buesser) [inline art/attacks/inference/membership_inference/black_box.py:77]: Could this base class ever be used for attacks on non-classification models?
```suggestion
        super().__init__(estimator=classifier)
```

File: art/attacks/attack.py [hunk 5]
```diff
@@ -334,12 +334,51 @@ def __init__(self, estimator, attack_feature: Union[int, slice] = 0):
...
+        super().set_params(**kwargs)
+        self._check_params()
+
+
+class MembershipInferenceAttack(InferenceAttack):
+    """
+    Abstract base class for membership inference attack classes.
+    """
+
+    def __init__(self, classifier: Union["CLASSIFIER_TYPE"]):
```
Comment (beat-buesser) [inline art/attacks/attack.py:360]: Could this base class ever be used for attacks on estimators other than for classification?
```suggestion
    def __init__(self, estimator: Union["CLASSIFIER_TYPE"]):
```

File: art/attacks/inference/membership_inference/black_box.py [hunk 6]
```diff
@@ -231,18 +226,22 @@ def fit(  # pylint: disable=W0613
                     loss.backward()
                     optimizer.step()
         else:
-            if self.attack_model_type == "gb":
-                y_ready = check_and_transform_label_format(y_new, len(np.unique(y_new)), return_one_hot=False)
-            else:
-                y_ready = check_and_transform_label_format(y_new, len(np.unique(y_new)), return_one_hot=True)
+            # if self.attack_model_type == "gb":
```
Comment (beat-buesser) [inline art/attacks/inference/membership_inference/black_box.py:229]: Was this change intended?

File: tests/attacks/inference/attribute_inference/test_meminf_based.py [hunk 7]
```diff
@@ -0,0 +1,356 @@
+# MIT License
+#
+# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2020
```
Comment (beat-buesser) [inline tests/attacks/inference/attribute_inference/test_meminf_based.py:3]:
```suggestion
# Copyright (C) The Adversarial Robustness Toolbox (ART) Authors 2021
```

File: art/attacks/inference/attribute_inference/meminf_based.py (hunk 1 above)
Comment (abigailgold) [reply art/attacks/inference/attribute_inference/meminf_based.py:40]: I know, this is the shortest I could come up with as a descriptive name. I could either make it non-descriptive, or shorten the long words (such as MemInf instead of MembershipInference) or remove the prefix AttributeInference (like you did for the label only attack). @beat-buesser What do you think?

File: art/attacks/inference/membership_inference/black_box.py [hunk 8]
```diff
@@ -243,7 +240,10 @@ def infer(self, x: np.ndarray, y: Optional[np.ndarray] = None, **kwargs) -> np.n
 
         :param x: Input records to attack.
         :param y: True labels for `x`.
-        :return: An array holding the inferred membership status, 1 indicates a member and 0 indicates non-member.
+        :param probabilities: a boolean indicating whether to return the predicted probabilities per class, or just
+                              the predicted class
+        :return: An array holding the inferred membership status, 1 indicates a member and 0 indicates non-member,
+                or class probabilities.
```
Comment (beat-buesser) [inline art/attacks/inference/membership_inference/black_box.py:246]:
```suggestion
        :return: An array holding the inferred membership status, 1 indicates a member and 0 indicates non-member,
                 or class probabilities.
```

File: art/attacks/inference/attribute_inference/meminf_based.py [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+                                  should support returning probabilities.
+        :param attack_feature: The index of the feature to be attacked or a slice representing multiple indexes in
+                               case of a one-hot encoded feature.
+        """
+        super().__init__(estimator=classifier, attack_feature=attack_feature)
+        if not all(t in type(classifier).__mro__ for t in membership_attack.estimator_requirements):
+            raise EstimatorError(membership_attack, membership_attack.estimator_requirements, classifier)
+
+        self.membership_attack = membership_attack
+
```
Comment (beat-buesser) [inline art/attacks/inference/attribute_inference/meminf_based.py:69]:
```suggestion
```

File: art/attacks/attack.py [hunk 10]
```diff
@@ -358,12 +366,49 @@ def __init__(self, estimator, attack_feature: Union[int, slice] = 0):
...
+        """
+        raise NotImplementedError
+
+    def set_params(self, **kwargs) -> None:
+        """
+        Take in a dictionary of parameters and applies attack-specific checks before saving them as attributes.
+        """
+        # Save attack-specific parameters
+        super().set_params(**kwargs)
+        self._check_params()
```
Comment (beat-buesser) [inline art/attacks/attack.py:384]: I think the same `set_params` is already inherited from `Attack`.
```suggestion
```

File: art/attacks/attack.py [hunk 11]
```diff
@@ -329,15 +329,23 @@ def __init__(self, estimator):
...
         """
         raise NotImplementedError
 
+    def set_params(self, **kwargs) -> None:
+        """
+        Take in a dictionary of parameters and applies attack-specific checks before saving them as attributes.
+        """
+        # Save attack-specific parameters
+        super().set_params(**kwargs)
+        self._check_params()
```
Comment (beat-buesser) [inline art/attacks/attack.py:347]: I think the same `set_params` is already inherited from `Attack`.
```suggestion
```

**Review by beat-buesser — APPROVED** [review]
Hi @abigailgold Thank you very much for the new inference attacks!
